opttrace/opttrace.go

```diff
@@ -0,0 +1,163 @@
+package opttrace
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strings"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+	"go.opentelemetry.io/otel/trace/noop"
+)
+
+const tracerName = "opttrace"
+
+var noopTracerProvider = noop.NewTracerProvider()
+
+// Tracer provides a tracing interface that generates traces only if configured
+// with a trace exporter
+type Tracer struct {
+	exportTP *sdktrace.TracerProvider
+}
+
+// Option provides Tracer configuration options
+type Option func(*config) error
+
+type config struct {
+	otlpEndpointURI string
+	sampler         sdktrace.Sampler
+	syncExport      bool
+	batchOpts       []sdktrace.BatchSpanProcessorOption
+	exporter        sdktrace.SpanExporter
+}
+
+// WithOTLPExporter configured an OTLP trace exporter
+func WithOTLPExporter(endpointURI string) Option {
+	return func(c *config) error {
+		c.otlpEndpointURI = endpointURI
+		return nil
+	}
+}
+
+// WithSampler sets the sampler to be used by the underlying tracing
+// provider. If not set, it takes the default of sampling based on whether the
+// parent span was sampled.
+func WithSampler(sampler sdktrace.Sampler) Option {
+	return func(c *config) error {
+		c.sampler = sampler
+		return nil
+	}
+}
+
+// WithBatchOptions allows overriding the default span batch processing options.
+func WithBatchOptions(opts []sdktrace.BatchSpanProcessorOption) Option {
+	return func(c *config) error {
+		c.batchOpts = opts
+		return nil
+	}
+}
+
+// WithSyncExport can be used in tests and disables batch span processing.
+func WithSyncExport() Option {
+	return func(c *config) error {
+		c.syncExport = true
+		return nil
+	}
+}
+
+// WithExporter sets a span exporter other than standard one provided by
+// WithOTLPExporter.
+func WithExporter(exp sdktrace.SpanExporter) Option {
+	return func(c *config) error {
+		c.exporter = exp
+		return nil
+	}
+}
+
+// New creates a Tracer that will create spans if configured with an exporter.
+// If not, the Span method will use a no-op tracing provider. When enabled, the
+// spans will have the supplied service name.  The context is used to initialize
+// a configured OTLP exporter, if any.
+func New(ctx context.Context, serviceName string, opts ...Option) (*Tracer, error) {
+	c := &config{}
+	for _, opt := range opts {
+		err := opt(c)
+		if err != nil {
+			return nil, err
+		}
+	}
+	var err error
+	exp := c.exporter
+	if exp == nil {
+		if c.otlpEndpointURI == "" {
+			return &Tracer{}, nil
+		}
+		exp, err = otlpExporter(ctx, c.otlpEndpointURI)
+		if err != nil {
+			return nil, err
+		}
+	}
+	resources, err := resource.New(context.Background(),
```

Comment:
should this take the passed in `ctx`?

Reply:
It probably could, but technically none of these resource detectors I'm using make use of it as far as I can see.  I think it's for future use.  I'll wire it up anyway.